Firestore/Source/Local/FSTLevelDBMigrations.mm

```diff
@@ -102,22 +145,37 @@ + (FSTLevelDBSchemaVersion)schemaVersionWithTransaction:
 }
 
 + (void)runMigrationsWithTransaction:(firebase::firestore::local::LevelDbTransaction *)transaction {
+  [self runMigrationsWithTransaction:transaction upToVersion:kSchemaVersion];
+}
+
++ (void)runMigrationsWithTransaction:(firebase::firestore::local::LevelDbTransaction *)transaction
+                         upToVersion:(FSTLevelDBSchemaVersion)upToVersion {
   FSTLevelDBSchemaVersion currentVersion = [self schemaVersionWithTransaction:transaction];
   // Each case in this switch statement intentionally falls through. This lets us
   // start at the current schema version and apply any migrations that have not yet
   // been applied, to bring us up to current, as defined by the kSchemaVersion constant.
   switch (currentVersion) {
     case 0:
+      if (upToVersion < 0) break;
       EnsureTargetGlobal(transaction);
       ABSL_FALLTHROUGH_INTENDED;
     case 1:
       // We're now guaranteed that the target global exists. We can safely add a count to it.
+      if (upToVersion < 1) break;
       AddTargetCount(transaction);
       ABSL_FALLTHROUGH_INTENDED;
-    default:
-      if (currentVersion < kSchemaVersion) {
-        SaveVersion(kSchemaVersion, transaction);
+    case 2:
+      if (upToVersion < 2) break;
+      if (currentVersion != 0) {
+        LevelDbSchema schema{transaction->database()};
```

Comment:
This highlights that maybe transactions are misnamed: they are batch writes that allow you to read your own writes, but they do not provide isolation. In fact, the production code contains asserts to prevent you from opening a new transaction while one is already outstanding. Given that you're going to be using multiple transactions over the course of `DropQueryCache`, I think it would be better to `Commit` the existing transaction and update the schema version first. Otherwise, imagine that the next schema migration fails and does not commit the transaction with schema version 1. We'll be a in state where we have applied `DropQueryCache`, but not prior migrations.

Reply:
Done.